src/pg_query_summary_truncate.c

```diff
@@ -0,0 +1,673 @@
+					add_truncation(state,
+							TRUNCATION_VALUES_LISTS,
+							node,
+							select_values_lists_len(stmt->valuesLists));
+
+				break;
+			}
+
+		case T_UpdateStmt:
+			{
+				UpdateStmt *stmt = castNode(UpdateStmt, node);
+
+				if (stmt->targetList != NULL)
+					add_truncation(state,
+							TRUNCATION_TARGET_LIST,
+							node,
+							update_target_list_len(stmt->targetList));
+
+				add_truncation_where_clause(state, node, stmt->whereClause);
+
+				break;
+			}
+
+		case T_DeleteStmt:
+			{
+				DeleteStmt *stmt = castNode(DeleteStmt, node);
+
+				add_truncation_where_clause(state, node, stmt->whereClause);
+
+				break;
+			}
+
+		case T_CopyStmt:
+			{
+				CopyStmt *stmt = castNode(CopyStmt, node);
+
+				add_truncation_where_clause(state, node, stmt->whereClause);
+
+				break;
+			}
+
+		case T_InsertStmt:
+			{
+				InsertStmt *stmt = castNode(InsertStmt, node);
+
+				if (stmt->cols != NULL)
+					add_truncation(state, TRUNCATION_COLS, node, cols_len(stmt->cols));
+
+				break;
+			}
+
+		case T_IndexStmt:
+			{
+				IndexStmt *stmt = castNode(IndexStmt, node);
+				add_truncation_where_clause(state, node, stmt->whereClause);
+				break;
+			}
+
+		case T_RuleStmt:
+			{
+				RuleStmt *stmt = castNode(RuleStmt, node);
+				add_truncation_where_clause(state, node, stmt->whereClause);
+				break;
+			}
+
+		case T_CommonTableExpr:
+			{
+				CommonTableExpr *stmt = castNode(CommonTableExpr, node);
+
+				if (stmt->ctequery != NULL) {
+					PgQueryDeparseResult result = pg_query_deparse(node);
+
+					if (result.error) {
+						state->error = result.error;
+						return false;
+					}
+
+					int32_t length = strlen(result.query);
+
+					add_truncation(state,
+							TRUNCATION_CTE_QUERY,
+							node,
+							length);
+				}
+
+				break;
+			}
+
+		case T_InferClause:
+			{
+				InferClause *stmt = castNode(InferClause, node);
+				add_truncation_where_clause(state, node, stmt->whereClause);
+				break;
+			}
+
+		case T_OnConflictClause:
+			{
+				OnConflictClause *stmt = castNode(OnConflictClause, node);
+
+				if (stmt->targetList != NULL)
+					add_truncation(state,
+							TRUNCATION_TARGET_LIST,
+							node,
+							update_target_list_len(stmt->targetList));
+
+				add_truncation_where_clause(state, node, stmt->whereClause);
+
+				break;
+			}
+
+		default:
+			break;
+	}
+
+	if (!pg_query_raw_tree_walker_supports(node))
+		return false;
+
+	state->depth++;
+
+	return raw_expression_tree_walker(node, summary_truncation_options, (void *) state);
+}
+
+static int cmp_possible_truncation_depth(const ListCell *a, const ListCell *b)
+{
+	const PossibleTruncation *pt_a = lfirst(a);
+	const PossibleTruncation *pt_b = lfirst(b);
+
+	return pt_a->depth - pt_b->depth;
+}
+
+static void expand_ellipses(char *str)
+{
+	// https://www.compart.com/en/unicode/U+2026
+	// pattern is a double quote, ellipses, and another double quote.
+	char pattern[5] = {'"', 0xE2, 0x80, 0xA6, '"'};
+
+	for (size_t i = 0; i < strlen(str); i++) {
```

Comment:
I think we'd need to subtract the pattern length from strlen here, so we don't overrun when the pattern would no longer fit into the remaining string.